theano/tensor/opt.py

```diff
@@ -5210,11 +5210,12 @@ def local_opt_alloc(node):
                     val = get_scalar_constant_value(input)
                     assert val.size == 1
                     # check which type of op
+                    casted = T.mul(*shapes).astype(str(val.dtype))
```

Comment:
`get_scalar_constant_value` digs through `cast` operations.
You could use `input.dtype`, though.